File: BlogApp.Node.API/src/comments/comments.controller.ts
import { Controller, Get, Post, Body, Put, Param, Delete, ParseIntPipe, BadRequestException, HttpException, Query } from '@nestjs/common';
import { CommentsService } from '../services/comments.service';
import { CommentDto } from '../models/dtos/comment.dto';
import { HttpStatus } from '@nestjs/common';
import { AuthService } from '../services/auth.service';

@Controller('comments')
export class CommentsController {

    constructor(
      private commentsService: CommentsService,
      private authService: AuthService) {}

  @Post()
  async create(@Body() input: CommentDto) {
      input.date = new Date().toString();
      input.userID = this.authService.user.id;
      const createdComm = await this.commentsService.create(input);
      if (createdComm == undefined){
        throw new HttpException('Bad request.', HttpStatus.BAD_REQUEST);
      }
  }

  @Get()
  async getTopComments(@Query() query: {postid: number, search: string, page: number}) {
    const comments = await this.commentsService.getTopComments(query.postid, query.page, query.search);
    return comments;
  }

  @Get(':id')
  async getReplies(
    @Param('id', ParseIntPipe) id: number,
    @Query() query: {postID: number, page: number
    }) {
    const replies = await this.commentsService.getReplies(query.postID, id, query.page);
    return replies;
  }


  @Put(':id')
  async update(@Param('id', ParseIntPipe) id: number, @Body() input: CommentDto) {
    if (id != input.id){
      throw new HttpException('Bad Request', HttpStatus.BAD_REQUEST);
    }

    const comment = await this.commentsService.put(input);
    if (comment == undefined){
      throw new HttpException('There is no comment with the given id.', HttpStatus.NOT_FOUND)
    }

    if (!this.authService.isOwner(comment.userID)){
      throw new HttpException('Cant touch this.', HttpStatus.FORBIDDEN);
    }
  }

  @Delete(':id')
  async remove(@Param('id', ParseIntPipe) id: number) {
     const comment = await this.commentsService.delete(id);
    if (comment == undefined){
      throw new HttpException('There is no comment with the given id.', HttpStatus.NOT_FOUND)
    }

    if (!this.authService.isOwner(comment.userID)){
      throw new HttpException('Cant touch this.', HttpStatus.FORBIDDEN);
    }
  }
}

File: BlogApp.Dotnet.WebAPI/Controllers/CommentsController.cs
using BlogApp.Dotnet.ApplicationCore.DTOs;
using BlogApp.Dotnet.ApplicationCore.Interfaces;
using BlogApp.Dotnet.ApplicationCore.Models;
using IdentityModel;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Serilog;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace BlogApp.Dotnet.API.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class CommentsController : ApiBaseController<PostsController>
    {
        private readonly ICommentsService _commentsService;

        public CommentsController(ICommentsService commentsService)
        {
            _commentsService = commentsService;
        }

        // GET: api/Comments
        [HttpGet]
        public PaginatedDTO<CommentsDTO> GetComments(int postID, string search, int page = 1)
        {
            var commentDTOs = _commentsService.GetTopComments(postID, page, search ??= "");
            Log.Information("API: All Paginated Comments retrieved from database.");

            return commentDTOs;
        }

        // GET: api/Comments/5
        [HttpGet("{id}")]
        public PaginatedDTO<CommentsDTO> GetReplies(int id, int postID, int page = 1)
        {
            var replyDTO = _commentsService.GetReplies(postID, id, page);
            return replyDTO;
        }

        // PUT: api/Comments/5
        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
        [HttpPut("{id}")]
        [Authorize]
        public async Task<IActionResult> PutComment(int id, [Bind("ID,UserID,Date,PostID,Content,ParentID")] CommentsDTO comment)
        {
            if (!IsAuthorized(comment.UserID))
            {
                return Forbid();
            }

            if (id != comment.ID || !ModelState.IsValid)
            {
                return BadRequest();
            }

            try
            {
                await _commentsService.Update(comment);
            }
            catch (DbUpdateConcurrencyException)
            {
                if (!(await CommentExists(comment)))
                {
                    return NotFound();
                }
                else
                {
                    throw;
                }
            }

            Log.Information("API: The comment has been updated.");
            return NoContent();
        }

        // POST: api/Comments
        // To protect from overposting attacks, see https://go.microsoft.com/fwlink/?linkid=2123754
        [HttpPost]
        [Authorize]
        public async Task<ActionResult<Comment>> PostComment([Bind("ID,PostID,Content,ParentID")] CommentsDTO comment)
        {
            comment.UserID = HttpContext.User.Claims.First(c => c.Type == JwtClaimTypes.Id).Value;
            comment.Date = DateTime.Now;

            ModelState.Remove("UserID");
            if (ModelState.IsValid)
            {
                await _commentsService.Add(comment);
                Log.Information("API: The comment has been added to database.");
                return NoContent();
            }

            return BadRequest();
        }

        // DELETE: api/Comments/5
        [HttpDelete("{id}")]
        [Authorize]
        public async Task<IActionResult> DeleteComment(int id)
        {
            var comment = await _commentsService.Get(id);

            if (comment == null)
            {
                return NotFound();
            }

            if (!IsAuthorized(comment.UserID))
            {
                return Forbid();
            }

            await _commentsService.Remove(comment.ID);

            Log.Information("API: The comment has been deleted.");
            return NoContent();
        }

        private async Task<bool> CommentExists(CommentsDTO comment)
        {
            return await _commentsService.Add(comment) != null;
        }
    }
}
